src/backend/distributed/executor/local_executor.c

```diff
@@ -123,53 +133,124 @@ ExecuteLocalTaskList(CitusScanState *scanState, List *taskList)
 {
 	EState *executorState = ScanStateGetExecutorState(scanState);
 	ParamListInfo paramListInfo = copyParamList(executorState->es_param_list_info);
-	int numParams = 0;
-	Oid *parameterTypes = NULL;
 	ListCell *taskCell = NULL;
 	uint64 totalRowsProcessed = 0;
+	Job *workerJob = scanState->distributedPlan->workerJob;
 
-	if (paramListInfo != NULL)
+	foreach(taskCell, taskList)
 	{
-		const char **parameterValues = NULL; /* not used anywhere, so decleare here */
+		Task *task = (Task *) lfirst(taskCell);
 
-		ExtractParametersForLocalExecution(paramListInfo, &parameterTypes,
-										   &parameterValues);
+		PlannedStmt *localPlan =
+			LocalTaskPlannedStmt(workerJob->jobQuery, task, paramListInfo);
 
-		numParams = paramListInfo->numParams;
+		LogLocalCommand(workerJob, task);
+
+		totalRowsProcessed +=
+			ExecuteLocalTaskPlan(scanState, localPlan, task->queryString);
 	}
 
-	foreach(taskCell, taskList)
+	return totalRowsProcessed;
+}
+
+
+/*
+ * LocalTaskPlannedStmt creates the PlannedStmt for the input task.
+ *
+ * The function behaves differently when the task->queryString is avaliable
+ * or not. See the comments in the function for the details.
+ */
+static PlannedStmt *
+LocalTaskPlannedStmt(Query *workerJobQuery, Task *task, ParamListInfo boundParams)
+{
+	Query *shardQuery = copyObject(workerJobQuery);
+
+	/*
+	 * For performance reasons, the queryString is not generated for
+	 * local-fast path queries. Instead, we simply update the shard
+	 * references to the actual shard relations.
+	 *
+	 * Otherwise, like multi-shard queries going through the local execution,
+	 * we'd need to parse the queryString back to Query before passing it
+	 * to the planner,
+	 */
+	if (task->queryString == NULL)
 	{
-		Task *task = (Task *) lfirst(taskCell);
+		ReplaceShardReferencesWalker((Node *) shardQuery, task);
```

Comment:
Is there a way to skip a few more steps here and replace the relation ID of the distributed table with that of the shard without doing the intermediate step of wrapping information in a function call like we do when joining reference tables with local tables? We're now in the unique situation that we can actually deparse the shard names without magic, since they are local tables.

(not sure, just asking)